File: decoder/frame/slice.go
package frame

import (
	"encoding/base64"
	"encoding/binary"
	"encoding/json"
	"fmt"
	"strconv"

	"github.com/w6xian/sloth/v3/internal/utils"
)

const TextMessage byte = 0x01
const BinaryMessage byte = 0x02
const LongMessage byte = 0x80
const CRC byte = 0x40

type DataSlice struct {
	// P 消息类型 6位有效 高2位为扩展位 0x80 2字节长度 0x40 校验crc
	P byte `json:"p"`
	// Name 分片名称（用于标识是哪个信息）
	N string `json:"n"` // len 2
	// Total 分片总数
	T byte `json:"t"`
	// Index 当前分片索引
	I byte `json:"i"`
	// Size 消息体总大小
	S uint32 `json:"s"`
	// Data 分片数据
	D []byte `json:"d"`
}

// Bytes 编码为 JSON 文本帧。
// 原实现走 json.Marshal（反射）：每条消息的每个分片都要付出反射代价。
// 手写编码器输出字节与 json.Marshal 完全等价（解码端 frame.FromType 用 json.Unmarshal，无需改动）。
func (s *DataSlice) Bytes() []byte {
	return s.AppendJSON(make([]byte, 0, s.jsonSize()))
}

// jsonSize 预估 JSON 编码长度（含少量余量），用于一次性分配。
func (s *DataSlice) jsonSize() int {
	// {"p":0,"n":"xx","t":0,"i":0,"s":0,"d":"base64"}
	return 64 + base64.StdEncoding.EncodedLen(len(s.D))
}

// AppendJSON 手写编码 DataSlice，追加到 dst。
func (s *DataSlice) AppendJSON(dst []byte) []byte {
	return AppendSliceJSON(dst, *s)
}

// AppendSliceJSON 是 DataSlice 的手写 JSON 编码器（值传递，便于栈上构造分片）。
func AppendSliceJSON(dst []byte, s DataSlice) []byte {
	dst = append(dst, '{')
	dst = append(dst, '"', 'p', '"', ':')
	dst = strconv.AppendInt(dst, int64(s.P), 10)
	dst = append(dst, ',', '"', 'n', '"', ':')
	dst = appendJSONStr(dst, s.N)
	dst = append(dst, ',', '"', 't', '"', ':')
	dst = strconv.AppendInt(dst, int64(s.T), 10)
	dst = append(dst, ',', '"', 'i', '"', ':')
	dst = strconv.AppendInt(dst, int64(s.I), 10)
	dst = append(dst, ',', '"', 's', '"', ':')
	dst = strconv.AppendInt(dst, int64(s.S), 10)
	dst = append(dst, ',', '"', 'd', '"', ':')
	dst = appendJSONBytes(dst, s.D)
	dst = append(dst, '}')
	return dst
}

// 与 encoding/json 字节级等价的转义/base64 编码实现复用 internal/utils，
// 避免与 message 包的实现出现漂移。
func appendJSONStr(dst []byte, s string) []byte {
	return utils.AppendJSONStr(dst, s)
}

func appendJSONBytes(dst []byte, b []byte) []byte {
	return utils.AppendJSONBytes(dst, b)
}

func (s *DataSlice) MuskCheck() byte {
	return s.P & 0x40
}

func (s *DataSlice) Encode() []byte {
	return Encode(s)
}

func get_header_size(lLen byte, checkCRC bool) byte {
	c := byte(0x02)
	if !checkCRC {
		c = 0
	}
	return lLen + 1 + 2 + 1 + 1 + c
}

func Encode(s *DataSlice, opts ...FrameOption) []byte {
	opt := newOption(opts...)
	// 1byte type
	// 2byte name
	// 1byte slices
	// 1byte index
	// 2/4byte size
	// 0/2byte crc
	// nbyte data
	tag := s.P & 0x3F
	checkCRC := s.P&CRC == CRC
	l := len(s.D)
	// 根据长度大小判断是否需要扩展tag
	if l <= 0xFFFF {
		opt.LengthSize = 2
	} else {
		tag |= LongMessage
		opt.LengthSize = 4
	}
	if opt.CheckCRC || checkCRC {
		checkCRC = true
		tag |= CRC
	}
	// 根据长度大小判断是否需要扩展tag
	// 1+2+1+1+2/4+[2]+len(s.D)
	headerSize := get_header_size(opt.LengthSize, checkCRC)

	buf := make([]byte, int(headerSize)+len(s.D))
	buf[0] = tag
	var name [2]byte
	raw := []byte(s.N)
	switch len(raw) {
	case 0:
		name[0] = '0'
		name[1] = '0'
	case 1:
		name[0] = '0'
		name[1] = raw[0]
	default:
		name[0] = raw[0]
		name[1] = raw[1]
	}
	buf[1] = name[0]
	buf[2] = name[1]
	buf[3] = byte(s.T)
	buf[4] = byte(s.I)
	lb := make([]byte, opt.LengthSize)
	if opt.LengthSize == 2 {
		binary.BigEndian.PutUint16(lb, uint16(len(s.D)))
	} else {
		binary.BigEndian.PutUint32(lb, uint32(len(s.D)))
	}
	if checkCRC {
		copy(buf[5:headerSize-2], lb)
		crc := utils.GetCrC(s.D)
		buf[headerSize-2] = crc[0]
		buf[headerSize-1] = crc[1]
	} else {
		copy(buf[5:headerSize], lb)
	}
	copy(buf[headerSize:], s.D)
	return buf
}

// DecodeSlice 从二进制数据中解码分片
func Decode(b []byte) (*DataSlice, error) {
	headerSize := get_header_size(2, false)
	if len(b) < int(headerSize) {
		return nil, fmt.Errorf("invalid slice data length")
	}
	tag := b[0]
	opt := newOption()
	if tag&LongMessage == 0 {
		opt.LengthSize = 2
	} else {
		opt.LengthSize = 4
	}
	if tag&CRC == CRC {
		opt.CheckCRC = true
	}

	l := uint32(binary.BigEndian.Uint16(b[5:7]))
	if opt.LengthSize == 4 {
		l = binary.BigEndian.Uint32(b[5:9])
	}
	headerSize = get_header_size(opt.LengthSize, opt.CheckCRC)
	if len(b) < int(headerSize)+int(l) {
		return nil, fmt.Errorf("invalid slice data length")
	}
	s := &DataSlice{
		P: tag,
		N: string(b[1:3]),
		T: b[3],
		I: b[4],
	}
	s.S = l

	data := b[headerSize : int(headerSize)+int(l)]
	// 校验crc
	if opt.CheckCRC {
		crc := b[headerSize-2 : headerSize]
		if !utils.CheckCRC(data, crc) {
			return nil, fmt.Errorf("invalid slice crc")
		}
	}
	s.D = data
	return s, nil
}

func serialize(v any) []byte {
	b, err := json.Marshal(v)
	if err != nil {
		return []byte{}
	}
	return b
}

File: examples/ws/web/slice.js
// Constants
const TextMessage = 0x01;
const BinaryMessage = 0x02;
const LongMessage = 0x80;
const CRC = 0x40;

// DataSlice class
class DataSlice {
    constructor(p, n, t, i, s, d) {
        this.P = p; // Message type
        this.N = n; // Slice name (2 bytes)
        this.T = t; // Total slices
        this.I = i; // Current slice index
        this.S = s; // Total message size
        this.D = d; // Slice data
    }

    Bytes() {
        return serialize(this);
    }

    MuskCheck() {
        return this.P & 0x40;
    }

    Encode(opts = []) {
        return Encode(this, opts);
    }
}



function newOption(opts = []) {
    const opt = new Option();
    for (const o of opts) {
        o(opt);
    }
    return opt;
}

function get_header_size(lLen, checkCRC) {
    let c = 0x02;
    if (!checkCRC) {
        c = 0;
    }
    return lLen + 1 + 2 + 1 + 1 + c;
}


function serialize(v) {
    try {
        return new TextEncoder().encode(JSON.stringify(v));
    } catch (err) {
        return new Uint8Array(0);
    }
}

// Frame options
class Option {
    constructor() {
        this.CheckCRC = false;
        this.LengthSize = 2;
    }
}

function CheckCRC() {
    return function(opt) {
        opt.CheckCRC = true;
    };
}

function IsComplete(src, dst) {
    if (!src || !dst) return false;
    if (src.length !== 2 || dst.length !== 2) return false;
    return src[0] === dst[0] && src[1] === dst[1];
}

function CheckCRC(src, crc) {
    return IsComplete(getCRC(src), crc);
}

// Encode function
function Encode(s, opts = []) {
    const opt = newOption(opts);
    // 1byte type
    // 2byte name
    // 1byte slices
    // 1byte index
    // 2/4byte size
    // 0/2byte crc
    // nbyte data
    let tag = s.P & 0x3F;
    let checkCRC = (s.P & CRC) === CRC;
    const l = s.D.length;
    
    // Determine if long message tag is needed based on length
    if (l <= 0xFFFF) {
        opt.LengthSize = 2;
    } else {
        tag |= LongMessage;
        opt.LengthSize = 4;
    }
    
    if (opt.CheckCRC || checkCRC) {
        checkCRC = true;
        tag |= CRC;
    }

    const headerSize = get_header_size(opt.LengthSize, checkCRC);
    const buf = new Uint8Array(headerSize + l);
    
    buf[0] = tag;
    
    // Write name (2 bytes)
    const nameBytes = new TextEncoder().encode(s.N);
    buf[1] = nameBytes[0] || 0;
    buf[2] = nameBytes[1] || 0;
    
    buf[3] = s.T;
    buf[4] = s.I;
    
    // Write length
    const dv = new DataView(buf.buffer, buf.byteOffset, buf.byteLength);
    if (opt.LengthSize === 2) {
        dv.setUint16(5, l, false); // Big endian
    } else {
        dv.setUint32(5, l, false); // Big endian
    }
    
    // Write CRC if needed
    if (checkCRC) {
        const crc = getCRC(s.D);
        buf[headerSize - 2] = crc[0];
        buf[headerSize - 1] = crc[1];
    }
    
    // Write data
    buf.set(s.D, headerSize);
    
    return buf;
}

// Decode function
function Decode(b) {
    let headerSize = get_header_size(2, false);
    if (b.length < headerSize) {
        throw new Error("invalid slice data length");
    }
    
    const tag = b[0];
    const opt = new Option();
    
    if ((tag & LongMessage) === 0) {
        opt.LengthSize = 2;
    } else {
        opt.LengthSize = 4;
    }
    
    if ((tag & CRC) === CRC) {
        opt.CheckCRC = true;
    }
    
    const dv = new DataView(b.buffer, b.byteOffset, b.byteLength);
    let l = dv.getUint16(5, false); // Big endian
    if (opt.LengthSize === 4) {
        l = dv.getUint32(5, false); // Big endian
    }
    
    headerSize = get_header_size(opt.LengthSize, opt.CheckCRC);
    if (b.length < headerSize + l) {
        throw new Error("invalid slice data length");
    }
    
    const s = new DataSlice();
    s.P = tag;
    s.N = new TextDecoder().decode(b.subarray(1, 3));
    s.T = b[3];
    s.I = b[4];
    s.S = l;
    
    const data = b.subarray(headerSize, headerSize + l);
    
    // Verify CRC if needed
    if (opt.CheckCRC) {
        const crc = b.subarray(headerSize - 2, headerSize);
        if (!CheckCRC(data, crc)) {
            throw new Error("invalid slice crc");
        }
    }
    
    s.D = data;
    return s;
}

// Export everything
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        TextMessage,
        BinaryMessage,
        LongMessage,
        CRC,
        DataSlice,
        Encode,
        Decode,
        GetCrC,
        IsComplete,
        CheckCRC,
        newOption
    };
}